### PR TITLE
Allow marking of earlier offsets

### PR DESCRIPTION
**Pull request by jerrygb**
This commit is in alignment with the sarama library commits to reset offsets
https://github.com/Shopify/sarama/commit/b966238f31778386bce80c28a12fde729e01cab4#diff-3ca6d659defd100fe2de43adf2b8f41e
https://github.com/Shopify/sarama/commit/96fa1c837819761fc5c706010a9677aab15ca46d#diff-3ca6d659defd100fe2de43adf2b8f41e

### CODE REVIEW

**Comment by jerrygb** [comment]
I believe it might be a good idea to combine these with changes from @aravindvs

File: consumer.go [hunk 1]
```diff
@@ -161,6 +161,34 @@ func (c *Consumer) MarkOffsets(s *OffsetStash) {
...
 }
 
+// ResetOffsets marks the provided message as processed, alongside a metadata string
+// that represents the state of the partition consumer at that point in time. The
+// metadata string can be used by another consumer to restore that state, so it
+// can resume consumption.
+//
+// Difference between ResetOffset and MarkOffset is that it allows to rewind to an earlier offset
+func (c *Consumer) ResetOffset(msg *sarama.ConsumerMessage, metadata string) {
+	c.subs.Fetch(msg.Topic, msg.Partition).MarkOffset(msg.Offset+1, metadata)
```
Comment (dim) [inline consumer.go]: Shouldn't that call `.ResetOffset(...)`

**Comment by jerrygb** [comment]
@dim I have fixed as per the comment. You were right. Additionally, added some tests.

**Comment by jerrygb** [comment]
@dim reopening for travis rebuild, due to seemingly okay test
https://travis-ci.org/bsm/sarama-cluster/jobs/322296975

File: cluster_test.go [hunk 2]
```diff
@@ -21,6 +21,7 @@ var (
 	testKafkaRoot  = "kafka_2.12-1.0.0"
 	testKafkaAddrs = []string{"127.0.0.1:29092"}
 	testTopics     = []string{"topic-a", "topic-b"}
+	testTopicsReset     = "topic-c"
```
Comment (dim) [inline cluster_test.go]: pls use `gofmt` or `goimports`

File: cluster_test.go [hunk 3]
```diff
@@ -73,6 +74,10 @@ var _ = BeforeSuite(func() {
 		"-name", "kafkaServer", "kafka.Kafka",
 		testDataDir("server.properties"),
 	)
+	if _, err := os.Stat(testKafkaData); err == nil  {
```
Comment (dim) [inline cluster_test.go]: No need for that clause, instead just the below will do:

```
Expect(os.RemoveAll(testKafkaData)).To(Succeed())
```

File: cluster_test.go (hunk 2 above)
Comment (dim) [reply cluster_test.go]: also, no real need to declare this as global, it's only used once

File: consumer_test.go [hunk 4]
```diff
@@ -315,6 +340,94 @@ var _ = Describe("Consumer", func() {
 		Expect(uniques).To(HaveLen(15000))
 	})
 
+	It("should consume/commit/reset/resume", func() {
```
Comment (dim) [inline consumer_test.go]: so I am not really sure we need another fuzzing test for this. I don't see the added value. what you have done in https://github.com/bsm/sarama-cluster/pull/197/files#diff-e85d03976d9766c4d00ac176511d4f06R188 seems enough, please remove that test.

File: cluster_test.go (hunk 2 above)
Comment (jerrygb) [reply cluster_test.go]: True. Removed assuming we dont need another fuzzing test.

File: consumer_test.go (hunk 4 above)
Comment (jerrygb) [reply consumer_test.go]: Removed the test. This was meant to check that we do actually rewind and the first offset is actually the one we expect. The earlier fuzzing test actually doesnt deal specifically with start offsets, right? 

Rather it deals with consuming atleast fixed number of unique entries (partition/offset pairs)?

I have removed it as you advised. But, let me know if we need check the start offsets when we start consumption.

**Comment by jerrygb** [comment]
@dim fixed as per comments. 

Also removed the additional fuzzing test. This was meant to check that we do actually rewind and, whether the first offset consumed is actually the one we expected. I suppose the earlier  fuzzing test actually deal specifically with start offsets, right?

For now, I have removed it as you advised. But, let me know if you need to me add those back in.

**Pull request merged**